core/blockchain.go

```diff
@@ -1779,3 +1786,47 @@ func (bc *BlockChain) SubscribeLogsEvent(ch chan<- []*types.Log) event.Subscript
 func (bc *BlockChain) SubscribeBlockProcessingEvent(ch chan<- bool) event.Subscription {
 	return bc.scope.Track(bc.blockProcFeed.Subscribe(ch))
 }
+
+// GetAncestorHash return the hash of ancestor at the given number
+func (bc *BlockChain) GetAncestorHash(ref *types.Header, target uint64) common.Hash {
+	number := ref.Number.Uint64() - 1
+	hash := ref.ParentHash
+	if target == number {
+		return hash
+	}
+	if target > number {
+		// Should never happen
+		log.Error("Ancestor number must be <= descendant", "ancestor target", target, "descendant", number)
+		return common.Hash{}
+	}
+	// Iterate the chain until we hit the target or we hit a ancestor
+	// within the storage
+	for !bc.hashHistory.Contains(number, hash) {
+		header := bc.GetHeader(hash, number)
+		if header == nil {
+			return common.Hash{}
+		}
+		number = header.Number.Uint64() - 1
+		hash = header.ParentHash
+		if number == target {
+			return hash
+		}
+	}
+	// The hash storage has the right ancestor chain
+	if h, found := bc.hashHistory.Get(target); found {
+		return h
+	}
+	// The number is too old - continue iteration from the
+	// earliest point
+	number, hash = bc.hashHistory.Oldest()
+	// At this point we iterate the remaining headers via the chain
```

Comment:
This comment is the same as the one above, perhaps reword the above and delete this:

```
	// The number is too old - continue iteration from the
	// earliest point via the chain.
```